File: 2021/2/main.py
from os import DirEntry


def take_input():
    inputs = []
    with open("input.txt", "r") as f:
        for line in f:
            inputs.append(line.rstrip())
    return inputs


def get_direction(command):
    return command.split(' ')[0]


def get_magnitude(command):
    return int(command.split(' ')[1])


def sol_1(commands):
    horizontal = 0
    depth = 0
    for command in commands:
        direction, magnitude = get_direction(command), get_magnitude(command)
        if direction == "forward":
            horizontal += magnitude
        elif direction == "up":
            depth -= magnitude
        elif direction == "down":
            depth += magnitude
    return horizontal * depth


def sol_2(commands):
    aim = 0
    horizontal = 0
    depth = 0
    for command in commands:
        direction, magnitude = get_direction(command), get_magnitude(command)
        if direction == "forward":
            horizontal += magnitude
            depth += aim*magnitude
        elif direction == "up":
            aim -= magnitude
        elif direction == "down":
            aim += magnitude
    return horizontal * depth


def main():
    commands = take_input()
    print(sol_1(commands))
    print(sol_2(commands))


if __name__ == "__main__":
    main()

File: 2021/2/main.rb
class Command
  attr_reader :direction, :magnitude

  def initialize(command)
    @direction = command.split.[] 0
    @magnitude = command.split[1].to_i
  end
end

def take_input
  File.read("input.txt").split("\n").map! {|element| Command.new(element)}
end

def sol_1(commands)
  horizontal = 0
  depth = 0
  commands.each do |command|
    case command.direction
    when "forward"
      horizontal += command.magnitude
    when "up"
      depth -= command.magnitude
    when "down"
      depth += command.magnitude
    end
  end
  horizontal*depth
end

def sol_2(commands)
  horizontal = 0
  depth = 0
  aim = 0
  commands.each do |command|
    case command.direction
    when "forward"
      horizontal += command.magnitude
      depth += aim * command.magnitude
    when "up"
      aim -= command.magnitude
    when "down"
      aim += command.magnitude
    end
  end
  horizontal*depth
end

def main
  commands = take_input
  puts sol_1(commands)
  puts sol_2(commands)
end

main